src/uct/api/uct.h

```diff
@@ -881,13 +936,121 @@ struct uct_ep_params {
      */
     uct_sockaddr_priv_pack_callback_t sockaddr_pack_cb;
+
+    /**
+     * The connection manager object as created by @ref uct_cm_open.
+     * @a cm or @a iface field must be initialized but not both.
+     */
+    uct_cm_h                          cm;
+
+    /**
+     * Connection request what was passed to
+     * @ref uct_listener_conn_request_callback_t .
+     */
+    uct_conn_request_h                conn_request;
+
+    union {
+        /**
+         * Callback that will be invoked when the endpoint on the client side
+         * is being connected to the server by a connection manager @ref uct_cm_h .
+         */
+        uct_ep_client_connect_cb_t      client;
+
+        /**
+         * Callback that will be invoked when the endpoint on the server side
+         * is being connected to a client by a connection manager @ref uct_cm_h .
+         */
+        uct_ep_server_connect_cb_t      server;
+    } sockaddr_connect_cb;
+
+    /**
+     * Callback that will be invoked when the endpoint is disconnected.
+     */
+    uct_ep_sockaddr_disconnect_cb_t     disconnect_cb;
+};
+
+
+/**
+ * @ingroup UCT_RESOURCE
+ * @brief Connection manager attributes, capabilities and limitations.
+ */
+struct uct_cm_attr {
+    size_t      max_conn_priv;  /**< Max size of the connection manager's
+                                     private data used for connection
+                                     establishment with sockaddr. */
+};
+
+
+/**
+ * @ingroup UCT_RESOURCE
+ * @brief Parameters for creating a connection manager object @ref uct_cm_h by
+ * @ref uct_cm_open
+ */
+struct uct_cm_params {
+    /**
+     * Mask of valid fields in this structure, using bits from
+     * @ref uct_cm_params_field. Fields not specified by this mask
+     * will be ignored.
+     */
+    uint64_t                          field_mask;
+
+    /**
+     * Memory domain name, as returned from @ref uct_component_query,
+     * mandatory field.
+     */
+    const char                        *md_name;
+
+    /**
+     * UCT worker, mandatory field.
+     */
+    uct_worker_h                      worker;
+};
+
+
+/**
+ * @ingroup UCT_RESOURCE
+ * @brief Parameters for creating a listener object @ref uct_listener_h by
+ * @ref uct_listener_create
+ */
+struct uct_listener_params {
+    /**
+     * Mask of valid fields in this structure, using bits from
+     * @ref uct_listener_params_field. Fields not specified by this mask
+     * would be ignored.
+     */
+    uint64_t                             field_mask;
+
+    /**
+     * Connection manager, mandatory field.
+     */
+    uct_cm_h                             cm;
+
+    /**
+     * Socket address to listen on, a mandatory field.
+     */
+    ucs_sock_addr_t                      sockaddr;
+
+    /**
+     * Backlog of incoming connection requests.
+     */
+    int                                  backlog;
+
+    /**
+     * Callback for an incoming connection request on the server
```

Comment:
Callback function for handling incoming connection requests
BTW do we need async/sync flags for this callback?
can it be called from another thread?

Reply:
@yosefe yes, this callback can be called from another thread (async thread in rdmacm case for example). it's documented in the cb's description that it has to be thread safe.
need to add 'flags' to this cb args then and add new defines to the cm attr?
UCT_CM_FLAG_CB_ASYNC/SYNC (like we have for an iface)

Reply:
if we don't have specific reason to make this callback thread safe, no need to add (we can extend this in the future with field_mask)